include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl

```diff
@@ -3,150 +3,189 @@
 
 #include <irr/builtin/material_compiler/glsl/common.glsl>
 
-void instr_eval_execute(in instr_t instr, in vec3 L)
+void instr_eval_execute(in instr_t instr, in MC_precomputed_t precomp, inout irr_glsl_LightSample s, inout irr_glsl_AnisotropicMicrofacetCache _microfacet)
 {
 	uint op = instr_getOpcode(instr);
 
 	//speculative execution
 	bsdf_data_t bsdf_data = fetchBSDFDataForInstr(instr);
+	const mat2x3 ior = bsdf_data_decodeIoR(bsdf_data, op);
+	const float ior_scalar = dot(CIE_XYZ_Luma_Y_coeffs, ior[0]);
+	const mat2x3 ior2 = matrixCompMult(ior, ior);
 	params_t params = instr_getParameters(instr, bsdf_data);
 	float bxdf_eval_scalar_part;
 	uint ndf = instr_getNDF(instr);
 	float a = params_getAlpha(params);
 	float a2 = a*a;
-#ifndef ALL_ISOTROPIC_BXDFS
 	float ay = params_getAlphaV(params);
 	float ay2 = ay*ay;
-#endif
+	const vec3 refl = params_getReflectance(params);
 
-	float cosFactor = op_isBSDF(op) ? abs(currBSDFParams.isotropic.NdotL):max(currBSDFParams.isotropic.NdotL,0.0);
+	const bool is_bsdf = !op_isBRDF(op); //note it actually tells if op is BSDF or BUMPMAP or SET_GEOM_NORMAL (divergence reasons)
 
-	if (cosFactor>FLT_MIN && op_hasSpecular(op))
-	{
-#ifdef NDF_GGX
-#ifndef ONLY_ONE_NDF
-		if (ndf==NDF_GGX) {
+#ifndef NO_TWOSIDED
+	handleTwosided(instr, s, _microfacet);
 #endif
```

Comment:
is the proper twosided handling implemented yet? because with different instruction streams per face/side this function shouldn't exist yet

Reply:
thats why i didnt even bother with proper twosided handling